src/libertem/udf/base.py

```diff
@@ -57,26 +63,226 @@
 DeviceClass = Literal['cpu', 'cuda']
 UDFKwarg = Union[Any, AuxBufferWrapper]
 UDFKwargs = Dict[str, UDFKwarg]
+ExecutionPlan = Dict[ArrayBackend, Iterable["UDF"]]
 
 
 def _get_dtype(
     udfs: List["UDF"],
     dtype: "nt.DTypeLike",
-    corrections: Optional[CorrectionSet]
+    corrections: Optional[CorrectionSet],
+    array_backends: Iterable[ArrayBackend],
 ) -> "nt.DTypeLike":
     tmp_dtype: "nt.DTypeLike"
     if corrections is not None and corrections.have_corrections():
         tmp_dtype = np.result_type(np.float32, dtype)
     else:
         tmp_dtype = np.dtype(dtype)
     for udf in udfs:
-        tmp_dtype = np.result_type(
+        tmp_dtype = result_type(
             udf.get_preferred_input_dtype(),
-            tmp_dtype
+            tmp_dtype,
+            *array_backends,
         )
     return tmp_dtype
 
 
+class TileConverter:
+    def __init__(self, tile: DataTile):
+        self._tile = tile
+        # prime cache with identity
+        self._cache = {get_backend(tile.data): tile.data}
+
+    def get(self, backend):
```

Comment:
I had a bug occur when writing the example notebook where a BACKEND_ALL-supporting udf + a normal MemoryDataSet(dense_array) but with a CUDA-only context, caused the data to convert to from dense numpy cupyx.scipy.sparse.csr_matrix rather than dense cp.array as I would have expected. I can't reproduce it unfortunately, and wondered if this is explicitly handled anywhere ? For a big array sparsification could be quite costly (or even undefined).

Reply:
Hmm I'll have to check the conversion efficiency matrix. It will choose the conversion based on that, but afair there's also some logic to consider the array format and device class when creating the execution plan.

Nice catch, I'll try to reproduce and see what is going on! I agree that it is sketchy to convert from dense to sparse like that, and probably not what one would expect.

Reply:
I think I found the source of the confusion. A dataset indicates preference by the order of backends. The memory dataset only preferred its native type, meaning the remaining ones were in random order, without taking conversion efficiency into account. I am now including a change where it sorts the backends by a cost metric, which prefers same backend, same sparsity and same device class. Within these categories it sorts by the conversion cost of sparseconverter. This requires an update of sparseconverter, too, to have an interface to get the conversion cost.

Reply:
It SHOULD be resolved now. :-)